src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -461,36 +681,124 @@ bool ClusterDiscovery::runMainThread(std::function<void()> up_to_date_callback)
     while (!finished)
     {
         bool all_up_to_date = true;
-        auto & clusters = clusters_to_update->wait(5s, finished);
-        for (auto & [cluster_name, need_update] : clusters)
+
+        if (!multicluster_discovery_paths->empty())
         {
-            auto cluster_info_it = clusters_info.find(cluster_name);
-            if (cluster_info_it == clusters_info.end())
+            std::unordered_map<String, ClusterInfo> new_dynamic_clusters_info;
+            std::unordered_set<size_t> unchanged_roots;
+            findDynamicClusters(new_dynamic_clusters_info, &unchanged_roots);
+
+            std::unordered_set<String> clusters_to_insert;
+            std::unordered_set<String> clusters_to_remove;
+
+            for (const auto & [cluster_name, info] : clusters_info)
             {
-                LOG_ERROR(log, "Unknown cluster '{}'", cluster_name);
-                continue;
+                if (!info.zk_root_index)
```

Comment:
Is it actually possible? In which circumstances?

Reply:
"Static" clusters from local xml-config.
This (existed before) ability allows to discover nodes for cluster `static_cluster_name`:
```
<clickhouse>
    <remote_servers>
        <static_cluster_name>
            <discovery>
                <path>/clickhouse/discovery/static_cluster_name</path>
```

This (new from this PR) ability allows to discover clustres, and these clusters can be created and removed dynamically:
```
<clickhouse>
    <remote_servers>
        <dynamic_clusters> # Actually this name is not used, can be any
            <discovery>
                <observer />
                <multicluster_root_path>/clickhouse/discovery</multicluster_root_path>
```

Reply:
Right,
but does not !multicluster_discovery_paths->empty()
guarantee that it is a dynamic cluster?

Reply:
No, can be both static and dynamic in the same time.